### PR TITLE
WIP: Fix pinging and timeout management

### PR DESCRIPTION
**Pull request by turingmachinerepair**

### CODE REVIEW

File: src/smpp_socket.erl [hunk 1]
```diff
@@ -471,10 +475,15 @@ handle_pkt(#pdu{command_id = CmdID, sequence_number = Seq} = Pkt,
...
-            {ok, NextState, State1};
+            State2 = case State1 of
+                #{in_flight := []} -> maps:without([response_time], State1);
+                #{in_flight := _} -> set_request_timeout(State1)
+            end,
+            {ok, NextState, State2};
         false ->
             report_unexpected_response(Pkt, StateName, State),
-            {ok, StateName, State}
+            State2 = set_keepalive_timeout(State, esme),
```
Comment (arentrue) [inline src/smpp_socket.erl]: Is it really needed here: on _incoming_ packet?

File: src/smpp_socket.erl [hunk 2]
```diff
@@ -471,10 +475,15 @@ handle_pkt(#pdu{command_id = CmdID, sequence_number = Seq} = Pkt,
                             rate_limit_cooldown -> StateName;
                             _ -> bound
                         end,
-            {ok, NextState, State1};
+            State2 = case State1 of
+                #{in_flight := []} -> maps:without([response_time], State1);
+                #{in_flight := _} -> set_request_timeout(State1)
```
Comment (arentrue) [inline src/smpp_socket.erl]: I'd move this to additional clause of `set_request_timeout/1` as the latter one incapsulates the handling of `response_time` state variable. E.g
```erlang
set_request_timeout(#{in_flight := []} = State) ->
    aps:without([response_time], State);
set_request_timeout(#{in_flight := InFlight} = State) ->
    ...
```

File: src/smpp_socket.erl [hunk 3]
```diff
@@ -360,7 +362,7 @@ bound(cast, {send_req, Body, Sender, Time}, State) ->
         true ->
             ok = dequeue_reg(State),
             State1 = reply(State, {error, sending_expired}, current_time(), Sender),
-            keep_state(State1);
+            keep_state(State1, []);
```
Comment (arentrue) [inline src/smpp_socket.erl]: What for? This does not change anything, as:
```erlang
keep_state(State1) ->
    keep_state(State, []).
```

File: src/smpp_socket.erl [hunk 4]
```diff
@@ -375,7 +377,7 @@ bound(cast, stop, State) ->
     {stop, normal, State};
 bound(EventType, Msg, State) ->
     State1 = callback(handle_event, EventType, Msg, ?FUNCTION_NAME, State),
-    keep_state(State1).
+    keep_state(State1, []).
```
Comment (arentrue) [inline src/smpp_socket.erl]: Same question as above.

File: src/smpp_socket.erl (hunk 2 above)
Comment (turingmachinerepair) [reply src/smpp_socket.erl]: Done

File: src/smpp_socket.erl (hunk 3 above)
Comment (turingmachinerepair) [reply src/smpp_socket.erl]: WIP artifact, removed.

File: src/smpp_socket.erl (hunk 4 above)
Comment (turingmachinerepair) [reply src/smpp_socket.erl]: WIP artifact, removed.

File: src/smpp_socket.erl (hunk 1 above)
Comment (turingmachinerepair) [reply src/smpp_socket.erl]: WIP artifact, removed.

File: src/smpp_socket.erl [hunk 5]
```diff
@@ -471,7 +474,8 @@ handle_pkt(#pdu{command_id = CmdID, sequence_number = Seq} = Pkt,
                             rate_limit_cooldown -> StateName;
                             _ -> bound
                         end,
-            {ok, NextState, State1};
+            State2 = set_request_timeout(State1),
```
Comment (arentrue) [inline src/smpp_socket.erl:477]: It makes sense to move keepalive_timeout on smsc when receiving any valid response as well:
```suggestion
            State2 = set_request_timeout(State1),
            State3 = set_keepalive_timeout(State2, smsc),
            {ok, NextState, State3};
```

Comment (turingmachinerepair) [reply src/smpp_socket.erl:477]: Done

**Review by arentrue — APPROVED** [review]
Neat!

**Pull request merged**